src/peft/tuners/xlora/model.py

```diff
@@ -0,0 +1,357 @@
+            self.internal_xlora_scalings = xlora_scalings
+
+        model.register_forward_pre_hook(hook, with_kwargs=True, prepend=True)
+
+        self._freeze_all_adapters()
+
+        total_swapped, device = convert_layers_to_xlora(
+            model_peft,
+            self,
+            peft_config,
+        )
+
+        n_classes = len(peft_config.adapters)
+        xlora_classifier = XLoraClassifier(model_peft, peft_config, n_classes, total_swapped, device)
+
+        # Setup the model internal state
+        self.__dict__["internal_xlora_classifier"] = xlora_classifier
+        self.internal_xlora_scalings = None  # type: ignore
+
+    def _freeze_all_adapters(self):
+        self.eval()
+        if not self.xlora_config.use_trainable_adapters:
+            for name, param in self.named_parameters():
+                if "lora_" in name:
+                    param.requires_grad = False
+
+    def generate(self, *args, **kwargs):
+        # Rely on LoraModel.__getattr__
+        res = self.model.generate(*args, **kwargs)  # type: ignore
+        # TODO(EricLBuehler): Evaluate effectiveness and performance degradation
+        self._freeze_all_adapters()
+        return res
+
+    def _mark_only_adapters_as_trainable(self, model: nn.Module) -> None:
+        active_adapters = []
+        copy = self.active_adapters.copy()
+        for name in self.active_adapters:
+            if not isinstance(self.peft_config[name], XLoraConfig):
+                active_adapters.append(name)
+        self.active_adapter = active_adapters
+        super()._mark_only_adapters_as_trainable(model)
+
+        self.active_adapter = copy
+
+    def _save_pretrained_hook(
+        self,
+        save_directory: str,
+        safe_serialization: bool = True,
+        is_main_process: bool = True,
+        **kwargs: Any,
+    ) -> None:
+        classifier: XLoraClassifier = self.internal_xlora_classifier
+
+        conf = self.xlora_config.__dict__.copy()
+
+        # So that the adapters are unloadable and the user is forced to set them for from_pretrained
+        conf["adapters"] = None
+        with open(os.path.join(save_directory, "xlora_config.json"), "w") as f:
+            json.dump(conf, f)
+
+        if safe_serialization:
+            # https://github.com/huggingface/peft/blob/main/src/peft/peft_model.py#L223
+            if is_main_process and safe_serialization:
+                save_model(classifier, os.path.join(save_directory, "xlora_classifier.safetensors"))
+        elif is_main_process:
+            state_dict = classifier.state_dict()
+            torch.save(state_dict, os.path.join(save_directory, "xlora_classifier.pt"))
+
+    def inject_adapter(self, model: nn.Module, adapter_name: str):
+        # XLora only supports a single adapter. It wouldn't make sense to have multiple XLora adapters,
+        # but multiple Lora adapters are essentially a requirement.
+        if len(set(self.peft_config.keys()) | {adapter_name}) > 1:
```

Comment:
Oh, this doesn't work after all. When trying to run the test, this fails because XLoRA tries to load the individual LoRA adapters, so there is going to be more than 1. I think we need a different approach.

I wonder if we should change the structure of `XLoraModel` a bit and if that might help to simplify the issue. Right now, `XLoraModel` is a `LoraModel`. I wonder if we can make it so that `XLoraModel` instead is its own class that has a `LoraModel` as an attribute. That `LoraModel` will contain all the different `LoraConfig`s and their respective adapters.

During the initialization of this `XLoraModel`, after the `LoraModel` is already initialized, we would go through it and replace all its `LoraLayer`s by their corresponding `XLoraLayer`s.

Additionally, the `XLoraModel` would hold a reference to the `XLoraClassifier`. Maybe that could help to avoid the recursion error, and hence allow us to get rid of the custom saving/loading code. WDYT?